File: hw1/HomeWork1/Parser.cs
using System;

namespace HomeWork1
{
    public static class Parser
    {
          private const int WrongOperation = 1;
          private const int WrongArgLength = 2;
          private const int WrongArgFormatInt = 3;
        
        public static int ParseArguments(string[] args, out int val1, out CalculatorOperation operation, out int val2)
        {
            var isInt1 = int.TryParse(args[0], out val1);
            var isInt2 = int.TryParse(args[2], out val2);
            
            if (!TryParseOperationOrQuit(args[1], out operation)) return WrongOperation;
            if (!CheckArgLength(args.Length)) return WrongArgLength;
            if (!TryParseOrQuit(args[0], isInt1) || !TryParseOrQuit(args[2], isInt2)) return WrongArgFormatInt;
            
            return 0;
        }

        private static bool TryParseOperationOrQuit(string arg, out CalculatorOperation operation)
        {
            operation = arg switch
            {
                "+" => CalculatorOperation.Plus,
                "-" => CalculatorOperation.Minus,
                "*" => CalculatorOperation.Multiply,
                "/" => CalculatorOperation.Divide,
                _ => 0
            };

            return operation != 0;
        }

        private static bool CheckArgLength(int length)
        {
            if (length == 3) return true;
            Console.WriteLine($"The program requires 3 arguments, but was {length}");
            return false;
        }
        
        private static bool TryParseOrQuit(string arg, bool isInt)
        {
            if (isInt) return true;
            Console.WriteLine($"Value is not int: {arg}");
            return false;
        }
    }
}

File: Homework5/Parser.fs
module Homework5.Parser

open System
let TryParseOperationOrQuit (val1: decimal, op: string, val2: decimal) =
    match op with
        |"+" -> Ok (val1, CalculatorOperation.Plus, val2)
        | "-" -> Ok (val1, CalculatorOperation.Minus, val2)
        | "*" -> Ok (val1, CalculatorOperation.Multiply, val2)
        | "/" -> Ok (val1, CalculatorOperation.Divide, val2)
        | _ -> Error HundledExceptions.WrongOperation
    
    
type MaybeBuilder() =
    member b.Bind(x, foo) =
        match x with
        | Ok y -> foo y
        | Error n -> Error n
    member b.Return x = Ok x
let maybe = MaybeBuilder()

let CheckArgLength (args: string[]) =
    match args.Length with
    | 3 -> Ok args
    | _ -> Error HundledExceptions.WrongArgLength

let TryParseArgOrQuit (args: string[]) =
    try
        Ok (args.[0] |> decimal, args.[1], args.[2] |> decimal)
    with
        | _ -> Error HundledExceptions.WrongArgFormat

let ParseArguments(args: string[]) =
    maybe{
        let! temp = CheckArgLength args
        let! temp2 = TryParseArgOrQuit temp
        let! result = TryParseOperationOrQuit temp2
        return result
    }
   
   